File: admin-frontend/src/app/core/member.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { firstValueFrom } from 'rxjs';

export interface MemberRole { discord_role_id:string; role_name:string; role_position:number; role_color:number; }
export interface Member {
  discord_user_id:string; username:string; global_name:string|null; nickname:string|null; avatar_url:string|null;
  bot:boolean; pending:boolean; is_active:boolean; joined_at:string|null; left_at:string|null; last_activity_at:string|null;
  communication_disabled_until:string|null; presence_status:string; activity_type:string|null; activity_name:string|null;
  voice_channel_id:string|null; voice_channel_name:string|null; client_desktop:boolean; client_mobile:boolean; client_web:boolean; last_presence_at:string|null; admin_note:string|null; game_nickname:string|null; alliance_tag:string|null; leadership_rank:'R5'|'R4'|'member'|null; preferred_language:string|null; verification_status:'not_verified'|'pending'|'verified'|'rejected'|'expired'; verification_updated_at:string|null; tags:string[]; shieldnet_blocked:boolean;
  watchlisted:boolean; risk_level:'low'|'medium'|'high'|'critical'; review_due_at:string|null; review_reason:string|null; roles:MemberRole[];
}
export interface MemberList { items:Member[]; total:number; page:number; page_size:number; }
export interface MemberStats { total:number; humans:number; bots:number; pending:number; timed_out:number; blocked:number; active_24h:number; inactive_30d:number; watchlisted:number; high_risk:number; review_due:number; }

export interface MemberTimelineItem { id:string; kind:string; title:string; detail:string|null; status:string|null; severity:string|null; occurred_at:string; metadata:Record<string,unknown>; }
export interface MemberInspector { member:Member; summary:{open_cases:number;resolved_cases:number;appeals:number;evidence:number;actions:number;verification_requests:number}; permissions:string[]; verification:Array<Record<string,unknown>>; timeline:MemberTimelineItem[]; }

export interface MemberAction { id:string; action_type:string; status:string; result_message:string|null; attempt_count:number; payload:Record<string,unknown>; }

export interface MemberCase {
  id:string; guild_id:string; discord_user_id:string; title:string; category:'warning'|'spam'|'harassment'|'security'|'appeal'|'other';
  severity:'low'|'medium'|'high'|'critical'; status:'open'|'investigating'|'resolved'|'dismissed'; priority:'low'|'normal'|'high'|'urgent'; description:string|null; resolution:string|null;
  assigned_to:string|null; created_by:string|null; due_at:string|null; first_response_at:string|null; resolved_at:string|null; created_at:string; updated_at:string;
}

export interface CaseEvidence {
  id:string; guild_id:string; case_id:string; evidence_type:'link'|'screenshot'|'message'|'document'|'other'; title:string; source_url:string|null; notes:string|null; created_by:string|null; created_at:string;
}
export interface CaseAppeal {
  id:string; guild_id:string; case_id:string; status:'submitted'|'under_review'|'accepted'|'rejected'|'withdrawn'; statement:string; decision:string|null; submitted_by_name:string|null; reviewed_by:string|null; reviewed_at:string|null; created_by:string|null; created_at:string; updated_at:string;
}

@Injectable({providedIn:'root'})
export class MemberService {
  constructor(private readonly http:HttpClient) {}

  list(guildId:string, filters:Record<string,string|number|undefined>):Promise<MemberList> {
    let params = new HttpParams();
    Object.entries(filters).forEach(([key,value]) => { if (value !== undefined && value !== '') params = params.set(key, String(value)); });
    return firstValueFrom(this.http.get<MemberList>(`/api/v1/discord/guilds/${guildId}/members`, {params}));
  }
  stats(guildId:string):Promise<MemberStats> { return firstValueFrom(this.http.get<MemberStats>(`/api/v1/discord/guilds/${guildId}/members/stats`)); }
  detail(guildId:string,userId:string):Promise<Member> { return firstValueFrom(this.http.get<Member>(`/api/v1/discord/guilds/${guildId}/members/${userId}`)); }
  updateProfile(guildId:string,userId:string,payload:{game_nickname:string|null;alliance_tag:string|null;leadership_rank:string|null;preferred_language:string|null;verification_status:string;admin_note:string|null;tags:string[];watchlisted:boolean;risk_level:string;review_due_at:string|null;review_reason:string|null}):Promise<Member> {
    return firstValueFrom(this.http.patch<Member>(`/api/v1/discord/guilds/${guildId}/members/${userId}/profile`, payload));
  }
  inspector(guildId:string,userId:string):Promise<MemberInspector> { return firstValueFrom(this.http.get<MemberInspector>(`/api/v1/discord/guilds/${guildId}/members/${userId}/inspector`)); }
  actions(guildId:string,userId:string):Promise<MemberAction[]> { return firstValueFrom(this.http.get<MemberAction[]>(`/api/v1/discord/guilds/${guildId}/members/${userId}/actions`)); }
  cases(guildId:string,userId:string):Promise<MemberCase[]> { return firstValueFrom(this.http.get<MemberCase[]>(`/api/v1/discord/guilds/${guildId}/members/${userId}/cases`)); }
  createCase(guildId:string,userId:string,payload:Record<string,unknown>):Promise<MemberCase> { return firstValueFrom(this.http.post<MemberCase>(`/api/v1/discord/guilds/${guildId}/members/${userId}/cases`, payload)); }
  updateCase(guildId:string,userId:string,caseId:string,payload:Record<string,unknown>):Promise<MemberCase> { return firstValueFrom(this.http.patch<MemberCase>(`/api/v1/discord/guilds/${guildId}/members/${userId}/cases/${caseId}`, payload)); }
  evidence(guildId:string,userId:string,caseId:string):Promise<CaseEvidence[]> { return firstValueFrom(this.http.get<CaseEvidence[]>(`/api/v1/discord/guilds/${guildId}/members/${userId}/cases/${caseId}/evidence`)); }
  createEvidence(guildId:string,userId:string,caseId:string,payload:Record<string,unknown>):Promise<CaseEvidence> { return firstValueFrom(this.http.post<CaseEvidence>(`/api/v1/discord/guilds/${guildId}/members/${userId}/cases/${caseId}/evidence`, payload)); }
  deleteEvidence(guildId:string,userId:string,caseId:string,evidenceId:string):Promise<void> { return firstValueFrom(this.http.delete<void>(`/api/v1/discord/guilds/${guildId}/members/${userId}/cases/${caseId}/evidence/${evidenceId}`)); }
  appeals(guildId:string,userId:string,caseId:string):Promise<CaseAppeal[]> { return firstValueFrom(this.http.get<CaseAppeal[]>(`/api/v1/discord/guilds/${guildId}/members/${userId}/cases/${caseId}/appeals`)); }
  createAppeal(guildId:string,userId:string,caseId:string,payload:Record<string,unknown>):Promise<CaseAppeal> { return firstValueFrom(this.http.post<CaseAppeal>(`/api/v1/discord/guilds/${guildId}/members/${userId}/cases/${caseId}/appeals`, payload)); }
  updateAppeal(guildId:string,userId:string,caseId:string,appealId:string,payload:Record<string,unknown>):Promise<CaseAppeal> { return firstValueFrom(this.http.patch<CaseAppeal>(`/api/v1/discord/guilds/${guildId}/members/${userId}/cases/${caseId}/appeals/${appealId}`, payload)); }
}

File: backups/patch-001-stage1-real-plugin-platform-v8-20260721-091900/app/services/member_service.py
from __future__ import annotations
from datetime import UTC, datetime, timedelta
import uuid

from fastapi import HTTPException, status
from sqlalchemy import String, delete, func, or_, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.member_actions import MemberAction, MemberActionStatus
from app.models.members import DiscordMember, DiscordMemberRole
from app.schemas.members import (
    MemberBatchSyncRequest,
    MemberDetailResponse,
    MemberListResponse,
    MemberProfileUpdate,
    MemberResponse,
    MemberRoleResponse,
    MemberStatsResponse,
)


class MemberService:
    def __init__(self, session: AsyncSession) -> None:
        self.session = session

    async def batch_sync(self, guild_id: int, payload: MemberBatchSyncRequest) -> int:
        seen: set[int] = set()
        now = datetime.now(UTC)

        for item in payload.members:
            seen.add(item.discord_user_id)
            result = await self.session.execute(
                select(DiscordMember).where(
                    DiscordMember.guild_id == guild_id,
                    DiscordMember.discord_user_id == item.discord_user_id,
                )
            )
            member = result.scalar_one_or_none()
            if member is None:
                member = DiscordMember(guild_id=guild_id, discord_user_id=item.discord_user_id, username=item.username)
                self.session.add(member)
                await self.session.flush()

            member.username = item.username
            member.global_name = item.global_name
            member.nickname = item.nickname
            member.avatar_url = item.avatar_url
            member.bot = item.bot
            member.pending = item.pending
            member.is_active = True
            member.joined_at = item.joined_at
            member.left_at = None
            member.communication_disabled_until = item.communication_disabled_until
            member.presence_status = item.presence_status
            member.activity_type = item.activity_type
            member.activity_name = item.activity_name
            member.voice_channel_id = item.voice_channel_id
            member.voice_channel_name = item.voice_channel_name
            member.client_desktop = item.client_desktop
            member.client_mobile = item.client_mobile
            member.client_web = item.client_web
            member.last_presence_at = item.last_presence_at
            member.updated_at = now

            await self.session.execute(delete(DiscordMemberRole).where(DiscordMemberRole.member_id == member.id))
            for role in item.roles:
                self.session.add(DiscordMemberRole(
                    member_id=member.id, guild_id=guild_id,
                    discord_role_id=role.discord_role_id, role_name=role.role_name,
                    role_position=role.role_position, role_color=role.role_color,
                ))

        if payload.complete_snapshot:
            query = update(DiscordMember).where(DiscordMember.guild_id == guild_id, DiscordMember.is_active.is_(True))
            if seen:
                query = query.where(DiscordMember.discord_user_id.not_in(seen))
            await self.session.execute(query.values(is_active=False, left_at=now, updated_at=now))

        await self.session.commit()
        return len(payload.members)

    async def activity(self, guild_id: int, discord_user_id: int, at) -> None:
        await self.session.execute(
            update(DiscordMember)
            .where(DiscordMember.guild_id == guild_id, DiscordMember.discord_user_id == discord_user_id)
            .values(last_activity_at=at or datetime.now(UTC), updated_at=datetime.now(UTC))
        )
        await self.session.commit()

    async def left(self, guild_id: int, discord_user_id: int, at) -> None:
        await self.session.execute(
            update(DiscordMember)
            .where(DiscordMember.guild_id == guild_id, DiscordMember.discord_user_id == discord_user_id)
            .values(is_active=False, left_at=at or datetime.now(UTC), updated_at=datetime.now(UTC))
        )
        await self.session.commit()

    def _filters(self, guild_id: int, query: str | None, member_type: str, status_filter: str, role_id: int | None, tag: str | None):
        now = datetime.now(UTC)
        filters = [DiscordMember.guild_id == guild_id]
        if status_filter != "left":
            filters.append(DiscordMember.is_active.is_(True))
        else:
            filters.append(DiscordMember.is_active.is_(False))
        if query:
            pattern = f"%{query.strip()}%"
            filters.append(or_(
                DiscordMember.username.ilike(pattern), DiscordMember.global_name.ilike(pattern),
                DiscordMember.nickname.ilike(pattern), func.cast(DiscordMember.discord_user_id, String).ilike(pattern),
                DiscordMember.admin_note.ilike(pattern), DiscordMember.game_nickname.ilike(pattern),
                DiscordMember.alliance_tag.ilike(pattern), DiscordMember.preferred_language.ilike(pattern),
            ))
        if member_type == "human": filters.append(DiscordMember.bot.is_(False))
        if member_type == "bot": filters.append(DiscordMember.bot.is_(True))
        if status_filter == "pending": filters.append(DiscordMember.pending.is_(True))
        if status_filter == "timed_out": filters.append(DiscordMember.communication_disabled_until > now)
        if status_filter == "blocked": filters.append(DiscordMember.shieldnet_blocked.is_(True))
        if status_filter == "inactive": filters.append(or_(DiscordMember.last_activity_at.is_(None), DiscordMember.last_activity_at < now - timedelta(days=30)))
        if status_filter == "watchlist": filters.append(DiscordMember.watchlisted.is_(True))
        if status_filter == "review_due": filters.append(DiscordMember.review_due_at.is_not(None)); filters.append(DiscordMember.review_due_at <= now)
        if role_id is not None:
            filters.append(DiscordMember.id.in_(select(DiscordMemberRole.member_id).where(DiscordMemberRole.discord_role_id == role_id)))
        if tag:
            filters.append(DiscordMember.tags.any(tag.strip().lower()))
        return filters

    async def _serialize(self, member: DiscordMember, detail: bool = False):
        roles = (await self.session.execute(
            select(DiscordMemberRole).where(DiscordMemberRole.member_id == member.id).order_by(DiscordMemberRole.role_position.desc())
        )).scalars().all()
        data = dict(
            discord_user_id=str(member.discord_user_id), username=member.username, global_name=member.global_name,
            nickname=member.nickname, avatar_url=member.avatar_url, bot=member.bot, pending=member.pending,
            is_active=member.is_active, joined_at=member.joined_at, left_at=member.left_at,
            last_activity_at=member.last_activity_at, communication_disabled_until=member.communication_disabled_until,
            presence_status=member.presence_status, activity_type=member.activity_type, activity_name=member.activity_name,
            voice_channel_id=str(member.voice_channel_id) if member.voice_channel_id is not None else None, voice_channel_name=member.voice_channel_name,
            client_desktop=member.client_desktop, client_mobile=member.client_mobile, client_web=member.client_web,
            last_presence_at=member.last_presence_at,
            admin_note=member.admin_note, game_nickname=member.game_nickname, alliance_tag=member.alliance_tag,
            leadership_rank=member.leadership_rank, preferred_language=member.preferred_language,
            verification_status=member.verification_status, verification_updated_at=member.verification_updated_at,
            tags=member.tags or [], shieldnet_blocked=member.shieldnet_blocked,
            watchlisted=member.watchlisted, risk_level=member.risk_level, review_due_at=member.review_due_at, review_reason=member.review_reason,
            roles=[MemberRoleResponse(discord_role_id=str(r.discord_role_id), role_name=r.role_name, role_position=r.role_position, role_color=r.role_color) for r in roles],
        )
        if detail:
            return MemberDetailResponse(**data, created_at=member.created_at, updated_at=member.updated_at, profile_updated_at=member.profile_updated_at)
        return MemberResponse(**data)

    async def list(self, guild_id: int, query: str | None, page: int, page_size: int, member_type: str = "all", status_filter: str = "active", role_id: int | None = None, tag: str | None = None, sort: str = "activity") -> MemberListResponse:
        filters = self._filters(guild_id, query, member_type, status_filter, role_id, tag)
        total = int((await self.session.execute(select(func.count(DiscordMember.id)).where(*filters))).scalar_one())
        order = {
            "name": (DiscordMember.username.asc(),),
            "joined": (DiscordMember.joined_at.desc().nullslast(),),
            "oldest": (DiscordMember.joined_at.asc().nullsfirst(),),
        }.get(sort, (DiscordMember.last_activity_at.desc().nullslast(), DiscordMember.username.asc()))
        result = await self.session.execute(select(DiscordMember).where(*filters).order_by(*order).offset((page - 1) * page_size).limit(page_size))
        return MemberListResponse(items=[await self._serialize(m) for m in result.scalars().all()], total=total, page=page, page_size=page_size)

    async def get(self, guild_id: int, discord_user_id: int) -> MemberDetailResponse:
        member = (await self.session.execute(select(DiscordMember).where(DiscordMember.guild_id == guild_id, DiscordMember.discord_user_id == discord_user_id))).scalar_one_or_none()
        if member is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Member not found")
        return await self._serialize(member, detail=True)

    async def update_profile(self, guild_id: int, discord_user_id: int, user_id: uuid.UUID, payload: MemberProfileUpdate) -> MemberDetailResponse:
        member = (await self.session.execute(select(DiscordMember).where(DiscordMember.guild_id == guild_id, DiscordMember.discord_user_id == discord_user_id))).scalar_one_or_none()
        if member is None:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Member not found")
        member.game_nickname = payload.game_nickname.strip() if payload.game_nickname else None
        member.alliance_tag = payload.alliance_tag.strip().upper() if payload.alliance_tag else None
        member.leadership_rank = payload.leadership_rank
        member.preferred_language = payload.preferred_language.strip().lower() if payload.preferred_language else None
        if member.verification_status != payload.verification_status:
            member.verification_status = payload.verification_status
            member.verification_updated_at = datetime.now(UTC)
        member.admin_note = payload.admin_note.strip() if payload.admin_note else None
        member.tags = payload.tags
        member.watchlisted = payload.watchlisted
        member.risk_level = payload.risk_level
        member.review_due_at = payload.review_due_at
        member.review_reason = payload.review_reason.strip() if payload.review_reason else None
        member.profile_updated_by = user_id
        member.profile_updated_at = datetime.now(UTC)
        await self.session.commit()
        await self.session.refresh(member)
        return await self._serialize(member, detail=True)

    async def stats(self, guild_id: int) -> MemberStatsResponse:
        now = datetime.now(UTC)
        base = [DiscordMember.guild_id == guild_id, DiscordMember.is_active.is_(True)]
        async def count(*extra):
            return int((await self.session.execute(select(func.count(DiscordMember.id)).where(*base, *extra))).scalar_one())
        return MemberStatsResponse(
            total=await count(), humans=await count(DiscordMember.bot.is_(False)), bots=await count(DiscordMember.bot.is_(True)),
            pending=await count(DiscordMember.pending.is_(True)), timed_out=await count(DiscordMember.communication_disabled_until > now),
            blocked=await count(DiscordMember.shieldnet_blocked.is_(True)), active_24h=await count(DiscordMember.last_activity_at >= now - timedelta(days=1)),
            inactive_30d=await count(or_(DiscordMember.last_activity_at.is_(None), DiscordMember.last_activity_at < now - timedelta(days=30))),
            watchlisted=await count(DiscordMember.watchlisted.is_(True)),
            high_risk=await count(DiscordMember.risk_level.in_(["high", "critical"])),
            review_due=await count(DiscordMember.review_due_at.is_not(None), DiscordMember.review_due_at <= now),
        )

    async def action_history(self, guild_id: int, discord_user_id: int, limit: int = 50) -> list[MemberAction]:
        return (await self.session.execute(
            select(MemberAction).where(MemberAction.guild_id == guild_id, MemberAction.discord_user_id == discord_user_id)
            .order_by(MemberAction.created_at.desc()).limit(limit)
        )).scalars().all()
